src/child.rs

```diff
@@ -0,0 +1,176 @@
+//!   is happening. This is important for showing `cargo build`'s output, for
+//!   example.
+
+use error::Error;
+use failure;
+use slog::Logger;
+use std::{
+    io::{self, BufRead},
+    process,
+    sync::mpsc,
+    thread, time,
+};
+use PBAR;
+
+fn taking_too_long(since: time::Instant) -> bool {
+    since.elapsed() > time::Duration::from_millis(200)
+}
+
+enum Output<S: AsRef<str>> {
+    Stdout(S),
+    Stderr(S),
+}
+
+fn print_child_output<S>(out: Result<Output<S>, io::Error>, command_name: &str)
+where
+    S: AsRef<str>,
+{
+    let message = match out {
+        Ok(Output::Stdout(line)) => format!("{} (stdout): {}", command_name, line.as_ref()),
+        Ok(Output::Stderr(line)) => format!("{} (stderr): {}", command_name, line.as_ref()),
+        Err(e) => format!("error reading {} output: {}", command_name, e),
+    };
+    PBAR.message(&message);
+}
+
+fn handle_output<I, S>(
+    output: I,
+    logger: &Logger,
+    should_print: bool,
+    stdout: &mut String,
+    stderr: &mut String,
+    command_name: &str,
+) where
+    I: IntoIterator<Item = Result<Output<S>, io::Error>>,
+    S: AsRef<str>,
+{
+    for out in output {
+        match out {
+            Ok(Output::Stdout(ref line)) => {
+                let line = line.as_ref().trim_end();
+                info!(logger, "{} (stdout): {}", command_name, line);
+                stdout.push_str(line);
+            }
+            Ok(Output::Stderr(ref line)) => {
+                let line = line.as_ref().trim_end();
+                info!(logger, "{} (stderr): {}", command_name, line);
+                stderr.push_str(line);
+            }
+            Err(ref e) => {
+                warn!(logger, "error reading output: {}", e);
+            }
+        }
+        if should_print {
+            print_child_output(out, command_name);
+        }
+    }
+}
+
+/// Run the given command and return its stdout.
+///
+/// If the command takes "too long", then its stdout and stderr are also piped
+/// to our stdout and stderr so that the user has an idea of what is going on
+/// behind the scenes.
+pub fn run(
+    logger: &Logger,
+    mut command: process::Command,
+    command_name: &str,
+) -> Result<String, failure::Error> {
+    info!(logger, "Running {:?}", command);
+
+    let mut child = command
+        .stdout(process::Stdio::piped())
+        .stderr(process::Stdio::piped())
+        .spawn()?;
+
+    let since = time::Instant::now();
+
+    let stdout = io::BufReader::new(child.stdout.take().unwrap());
+    let stderr = io::BufReader::new(child.stderr.take().unwrap());
+
+    let (send, recv) = mpsc::channel();
+    let stdout_send = send.clone();
+    let stderr_send = send;
+
+    // Because pipes have a fixed-size buffer, we need to keep reading stdout
+    // and stderr on a separate thread to avoid potential dead locks with
+    // waiting on the child process.
+
+    let stdout_handle = thread::spawn(move || {
+        for line in stdout.lines() {
```

Comment:
I've had unfortunately bad experiences in the past line-buffering stdout. An example of the bad behavior is that libtest in quiet mode used to simply print a `.`-per-test and then at the very end it'd print a newline. This logic means that no output would happen until the very very end where everything comes out all at once. We [fixed this bug for Rust](https://github.com/alexcrichton/stamp-rs/pull/1) in the past, and I think a similar fix could be applied here where this just streams data as it comes, and the data processor specially handles newlines in the stream to append output.